src/product_config_utils.rs

```diff
@@ -0,0 +1,1034 @@
+/// merged with that of the `role` it's self.
+/// In addition, the `*_overrides` settings are also merged in the resulting configuration
+/// with the highest priority.
+/// The merge priority chain looks like this:
+///
+/// group overrides -> group config -> role overrides -> role config
+///
+/// where '->' means "overwrites if existing or adds".
+///
+/// The output is a map with one entry, keyed by `role_name` and the value is a map where all
+/// configuration properties defined in the `role` are partitioned by `PropertyNameKind`.
+/// # Arguments
+/// - `resource`       - Not used directly. It's passed on to the `Configuration::compute_*` calls.
+/// - `role_name`      - Used as key in the output and to partition the configuration properties.
+/// - `role`           - The role for which to transform the configuration parameters.
+/// - `property_kinds` - Used as "buckets" to partition the configuration properties by.
+///
+fn transform_role_to_config<T>(
+    resource: &T::Configurable,
+    role_name: &str,
+    role: &Role<T>,
+    property_kinds: &[PropertyNameKind],
+) -> HashMap<String, HashMap<PropertyNameKind, BTreeMap<String, String>>>
+where
+    T: Configuration,
+{
+    let mut result = HashMap::new();
+
+    let role_properties =
+        partition_properties_by_kind(resource, role_name, &role.config, property_kinds);
+
+    // for each role group ...
+    for (role_group_name, role_group) in &role.role_groups {
+        // ... compute the group properties ...
+        let role_group_properties = partition_properties_by_kind(
+            resource,
+            role_group_name,
+            &role_group.config,
+            property_kinds,
+        );
+
+        // ... and merge them with the role properties.
+        let mut role_properties_copy = role_properties.clone();
+        for (property_kind, properties) in role_group_properties {
+            role_properties_copy
+                .entry(property_kind)
+                .or_default()
+                .extend(properties);
+        }
+
+        result.insert(role_group_name.clone(), role_properties_copy);
+    }
+
+    result
+}
+
+/// Given a `config` object and the `property_kind` vector, it uses the `Configuration::compute_*` methods
+/// to partition the configuration properties by `PropertyNameKind`.
+///
+/// The output is map where the configuration properties are keyed by `PropertyNameKind`.
+///
+/// # Arguments
+/// - `resource`       - Not used directly. It's passed on to the `Configuration::compute_*` calls.
+/// - `name`           - Not used directly but passed on to the `Configuration::compute_*` calls. It usually
+///                      contains a role name or a role group name.
+/// - `config`         - The configuration properties to partition.
+/// - `property_kinds` - The "buckets" used to partition the configuration properties.
+///
+fn partition_properties_by_kind<T>(
```

Comment:
Why do you call this `partition` when the others are all called `transform`?
(but I'm fine with it)